docs/PerformanceTest.md

```diff
@@ -0,0 +1,136 @@
+# eBPF Performance Benchmark Design
+
+## Overview
+
+The **eBPF Performance Benchmark suite (Perf)** aims to evaluate the performance of the eBPF runtime, pinpoint areas for enhancement, and ensure timely detection and mitigation of performance regressions. The suite provides a comprehensive view of build-to-build performance, categorized by scenarios, aiding in the identification of areas requiring improvement.
+
+## Architecture
+
+The Perf suite comprises these primary components:
+
+1. **Test Runner**: An executable designed to load, execute, and record metrics for an eBPF program.
+2. **Test Suites**: A collection of eBPF programs that comprehensively test various aspects of the eBPF runtime.
+3. **CI/CD Workflow**: A set of YAML files and scripts responsible for invoking the Test Runner with Test Suites, uploading resulting data to a central repository, and creating issues in case of performance regressions.
+
+## Test Runner
+
+The Test Runner (Runner) is a C++ program utilizing the libbpf library. It loads an eBPF program (in the kernel) and employs bpf_prog_test_run_opts to execute the BPF program on specified CPUs for a specified number of iterations. Subsequently, it calculates and outputs the average duration of the BPF programs. As a stretch goal, the Runner is designed to be platform-agnostic, enabling it to target both Windows and Linux eBPF runtimes.
+
+## Test Suites
+
+Each test is defined by a YAML file outlining test parameters and one or more associated eBPF programs. The parameters encompass:
+
+1. **Program -> CPU Assignment**: Specifies which CPUs execute each program, allowing concurrent execution.
```

Comment:
Should this document recommend how many programs should be run in parallel? Noting that this is a perf test and not stress test; what is the motivation behind running multiple programs? Lock contention?

Reply:
Given that the BPF code is intended to be used in a highly concurrent fashion, it makes sense to measure it under those conditions.

Reply:
Marking as resolved as this is now addressed in the Test Runner section.